pkgs/pkgs-lib/formats.nix

```diff
@@ -69,6 +69,51 @@ rec {
 
   };
 
+  jsonWithRuntimeSubstitution = {}: rec {
+    inherit (json {}) type;
+
+    findSubstitutions = let
+      prefixSubstitution = attrOrIndex: {attrpath, source}: { attrpath = [attrOrIndex] ++ attrpath; inherit source; };
+    in value:
+        if null == value || lib.isBool value || lib.isInt value || lib.isFloat value || lib.isString value || lib.isPath value then []
+        else
+          if lib.isList value then
+            lib.concatLists (lib.imap0 (index: value: map (prefixSubstitution index) (findSubstitutions value)) value)
+          else if lib.isAttrs value then
+            if value._runtimeSource or null != null then
+              # We shouldn't have anything other than a runtime source if this is a runtime-sourced string.
+              assert lib.attrNames value == ["_runtimeSource"];
+                [{attrpath = []; source = value._runtimeSource;}]
+            else
+              lib.concatLists (lib.mapAttrsToList (name: value: map (prefixSubstitution name) (findSubstitutions value)) value)
+          else
+            throw "unexpected type for jsonWithRuntimeSubstitution. How did you get this past module type checking??"
+    ;
+
+    generate = filename: value: let
```

Comment:
I don't see how this would properly handle typos in the variable names but it could also be that I just didn't spot it because the code is a bit to complex for me.

Reply:
I'm afraid I'm not sure what you mean with that: the variables inside jq are generated with `imap0` (i.e. `file0`, `file1`, etc.) pointing to the files where the secrets are.
This was done on purpose to avoid having to build fancy, error-prone serialisation logic in case the filename as special chars that cannot be used for variable names.

Reply:
What happens when I reference a secret for which no mapping to a file exists? Or the oath to the file doesn't exist?